File: backend/src/main/java/com/alahadattars/service/ReviewService.java
package com.alahadattars.service;

import com.alahadattars.dto.review.ReportReviewRequest;
import com.alahadattars.dto.review.ReviewRequest;
import com.alahadattars.dto.review.ReviewResponse;
import com.alahadattars.dto.review.ReviewSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface ReviewService {

    // Customer Methods
    ReviewResponse createReview(String email, ReviewRequest request);
    ReviewResponse updateReview(String email, Long id, ReviewRequest request);
    void deleteReview(String email, Long id);
    
    Page<ReviewResponse> getProductReviews(Long productId, Integer rating, String email, Pageable pageable);
    ReviewSummaryResponse getProductReviewSummary(Long productId);
    
    ReviewResponse toggleHelpfulVote(String email, Long id);
    void reportReview(String email, Long id, ReportReviewRequest request);
    
    ReviewResponse uploadReviewImages(String email, Long id, List<MultipartFile> files);

    // Admin Methods
    Page<ReviewResponse> getAllReviews(String search, Pageable pageable);
    ReviewResponse toggleVisibility(Long id, boolean hide);
    ReviewResponse adminReply(Long id, String reply);
    void adminDeleteReview(Long id);
}

File: frontend/src/services/reviewService.ts
import { apiClient } from '../api/axios';
import type { ApiResponse, PaginatedResponse, Review, ReviewSummary, ReviewRequest, ReportReviewRequest } from '../types';

export const reviewService = {
  // --- Customer Endpoints ---
  
  getProductReviews: async (productId: number | string, params?: Record<string, any>): Promise<PaginatedResponse<Review>> => {
    const response = await apiClient.get<ApiResponse<PaginatedResponse<Review>>>(`/reviews/product/${productId}`, { params });
    return response.data.data!;
  },

  getProductReviewSummary: async (productId: number | string): Promise<ReviewSummary> => {
    const response = await apiClient.get<ApiResponse<ReviewSummary>>(`/reviews/product/${productId}/summary`);
    return response.data.data!;
  },

  createReview: async (data: ReviewRequest): Promise<Review> => {
    const response = await apiClient.post<ApiResponse<Review>>('/reviews', data);
    return response.data.data!;
  },

  updateReview: async (id: number | string, data: ReviewRequest): Promise<Review> => {
    const response = await apiClient.put<ApiResponse<Review>>(`/reviews/${id}`, data);
    return response.data.data!;
  },

  deleteReview: async (id: number | string): Promise<void> => {
    await apiClient.delete(`/reviews/${id}`);
  },

  uploadImages: async (id: number | string, images: File[]): Promise<Review> => {
    const formData = new FormData();
    images.forEach(img => formData.append('images', img));
    const response = await apiClient.post<ApiResponse<Review>>(`/reviews/${id}/images`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data.data!;
  },

  toggleHelpful: async (id: number | string): Promise<Review> => {
    const response = await apiClient.post<ApiResponse<Review>>(`/reviews/${id}/helpful`);
    return response.data.data!;
  },

  reportReview: async (id: number | string, data: ReportReviewRequest): Promise<void> => {
    await apiClient.post(`/reviews/${id}/report`, data);
  },

  // --- Admin Endpoints ---

  getAllReviews: async (params?: Record<string, any>): Promise<PaginatedResponse<Review>> => {
    const response = await apiClient.get<ApiResponse<PaginatedResponse<Review>>>('/admin/reviews', { params });
    return response.data.data!;
  },

  toggleVisibility: async (id: number | string, hide: boolean): Promise<Review> => {
    const response = await apiClient.patch<ApiResponse<Review>>(`/admin/reviews/${id}/visibility`, { hide });
    return response.data.data!;
  },

  adminReply: async (id: number | string, reply: string): Promise<Review> => {
    const response = await apiClient.patch<ApiResponse<Review>>(`/admin/reviews/${id}/reply`, { reply });
    return response.data.data!;
  },

  adminDeleteReview: async (id: number | string): Promise<void> => {
    await apiClient.delete(`/admin/reviews/${id}`);
  }
};
